src/ui/marker.js

```diff
@@ -268,16 +315,84 @@ class Marker {
             this._lngLat = smartWrap(this._lngLat, this._pos, this._map.transform);
         }
 
-        this._pos = this._map.project(this._lngLat)._add(this._offset);
+        const pos = this._pos = this._map.project(this._lngLat);
+
+        let anchor = Point.convert(this.options && this.options.anchor || 'middle');
+        const offset = normalizeOffset(this.options && this.options.offset || [0, 0]);
+
+
+        if (!anchor) {
+            const width = this._element.offsetWidth,
+                height = this._element.offsetHeight;
+
+            if (pos.y + offset.bottom.y < height) {
+                anchor = ['top'];
+            } else if (pos.y > this._map.transform.height - height) {
+                anchor = ['bottom'];
+            } else {
+                anchor = [];
+            }
+
+            if (pos.x < width / 2) {
+                anchor.push('left');
+            } else if (pos.x > this._map.transform.width - width / 2) {
+                anchor.push('right');
+            }
+
+            if (anchor.length === 0) {
+                anchor = 'middle';
+            } else {
+                anchor = anchor.join('-');
+            }
+        }
+
+        const offsetedPos = pos.add(offset[anchor]);
```

Comment:
for some reason, this part right here seems to be breaking all my tests, usually in the form of add `undefined` or expecting a `truthy value`